### ISSUE TITLE
Optimizer filtering out data which is not outlier

### ISSUE THREAD

**Issue by user1823** [open-spaced-repetition/fsrs-optimizer#16]
Summary: The issue is that the optimizer is filtering out a part of my data just because I have a large amount of data with delta_t = 1 (which is a result of my past Anki settings).

Detailed description:
The current optimizer is filtering out my data with first rating = 3 and delta_t = 4 (and above). But, the initial stability for Good for my collection is more than 10 days (exact value depends on which version of the optimizer I use). This means that the optimizer is not using my newer data for training (unless the count with high delta_t eventually becomes so large that it fits into the threshold).

<img width="501" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/0d96b23d-6e66-46e4-a70a-212b85b096f4">

You might say that there is not much data for delta_t > 10 in the above image. This is because of two reasons:
1. My request retention is 0.94. So, stability of 10 days means an interval of 5.74 days.
2. This data is from when I was using FSRS v3. There, initial stability for Good was likely somewhat smaller.

As I said before, the reason that a large number of my cards (3944) have the first interval = 1d (with first rating = 3) is that 1d was my second learning step when I was using the default Anki algorithm. But, this should not be a factor to decide how well the optimizer works for me.

_Originally posted by @user1823 in https://github.com/open-spaced-repetition/fsrs4anki/issues/348#issuecomment-1645515720_

**Comment by user1823** [comment]
@L-M-Sherlock, what do you think about the idea of using ln(delta_t) to calculate quartiles and IQR?

I remember that it "increased" the RMSE but that was an artifact in my opinion.

Probably, we should implement this without thinking about the RMSE. We indeed made several changes without even testing the effect on the RMSE.

Otherwise, we should find out a better way to verify if using ln(delta_t) to calculate quartiles and IQR is better than using just delta_t.

**Comment by user1823** [comment]
Here is my S0 dataset. In my opinion, the ones having a pink background can be considered outliers. But probably, even some more rows could be considered outliers.

<img width="400" alt="image" src="https://github.com/open-spaced-repetition/fsrs-optimizer/assets/92206575/1dbbe81e-0864-40fa-9efe-dc53d70a13c8">

The tsv file is here: 
[stability_for_pretrain.tsv.zip](https://github.com/open-spaced-repetition/fsrs-optimizer/files/12315007/stability_for_pretrain.tsv.zip) (remove .zip at the end)

**Comment by L-M-Sherlock** [comment]
> Here is my S0 dataset. In my opinion, the ones having a pink background can be considered outliers. But probably, even some more rows could be considered outliers.

![image](https://github.com/open-spaced-repetition/fsrs-optimizer/assets/32575846/04013c9b-c963-4dcb-84c3-714b115d1052)

I think we can filter these data via setting threshold in the `count`.

![image](https://github.com/open-spaced-repetition/fsrs-optimizer/assets/32575846/8b215d70-5e1a-49c1-af7a-da3b4b07b64f)

Median seems to be a good candidate here.

**Comment by user1823** [comment]
Sorry, but I couldn't understand anything in this comment. What is the graph showing and what do the axes represent? Also, what are the three values in the second image?

**Comment by Expertium** [comment]
I think Sherlock's idea is to filter outliers based on the number of reviews with a certain interval length. For example, if there have been 20 reviews with delta_t = 2 and 1 review with delta_t = 20, then the latter would be considered an outlier, but not because of the length of the interval but rather because there has only been one such review. That's the gist of it, unless I also misunderstood Sherlock.

**Comment by Expertium** [comment]
So if we filter out all data where count < median(count) on user1823's data, here's what it will look like:
![image](https://github.com/open-spaced-repetition/fsrs-optimizer/assets/83031600/f9d52778-1402-4ddb-bedb-8774206ed898)

**Comment by L-M-Sherlock** [comment]
I mean the method is based on the quantile of count. It has a problem: it will always filter out x% `delta_t`. So we should consider other methods.

**Comment by Expertium** [comment]
How about trying LOF again? I know I've been a little annoying about it, but this seems like a good usecase for it. We have 3 features: delta_t, y (mean) and count. Pass all three of them into LOF.

**Comment by Expertium** [comment]
Though it's still unclear how to test it without relying on RMSE.

**Comment by user1823** [comment]
We can try using LOF. But, in the absence of an objective measure to test it, we will have to look which delta_t were filtered out by this approach and then subjectively decide if this approach is better or not.

Otherwise, we should just trust Expertium and use ln(delta_t) to calculate quartiles and IQR. This way, we would be using a widely used formula (Q3 + 1.5 × IQR) to find the outliers instead of LOF (which can possibly act in unexpected ways for some collections). Though this approach "increased" the RMSE in our testing, it was probably just an artifact (I know that I have said this too many times).

**Comment by Expertium** [comment]
I think Sherlock's idea (as I explained it [here](https://github.com/open-spaced-repetition/fsrs-optimizer/issues/16#issuecomment-1674559472)) could also work. It will filter out **half of all intervals**, **not half of all reviews**. In fact, in that image it only filtered out 27 reviews (for "Good") out of 5476. If LOF fails for whatever reason, I suggest this.
@L-M-Sherlock I recommend trying Local Outlier Factor with the following three features: ln(count), ln(count) and y_mean. SO it will work with three-dimensional data. Please, do the usual test of statistical significance.

**Comment by user1823** [comment]
> I think Sherlock's idea (as I explained it here) could also work. It will filter out half of all intervals ...

But, this method will work only for this particular data. For example, apply the same method to `Easy` in my data and you will see what I mean.

**Comment by Expertium** [comment]
> For example, apply the same method to `Easy` in my data and you will see what I mean.

![image](https://github.com/open-spaced-repetition/fsrs-optimizer/assets/83031600/10412673-1dd7-469c-bb74-526ffed430b6)



> But, it still has a problem: It will always filter out some part of the data even if it should ideally not be considered an outlier.
> 
> The IQR method has an advantage that it doesn't filter out any data if the data is more homogenous.

The reason why I'm recommending LOF ofer the more simple approaches is that LOF can work with multidimensional data, in fact, it might work _better_ with multidimensional data.

**Comment by L-M-Sherlock** [comment]
> The reason why I'm recommending LOF over the more simple approaches is that LOF can work with multidimensional data, in fact, it might work _better_ with multidimensional data.

But you should understand the mechanism of LOF. It filters out the outliers based on density. In out case, the outliers could also be concentrated in a range of intervals with high density. I will test it tomorrow.

**Comment by L-M-Sherlock** [comment]
> The reason why I'm recommending LOF over the more simple approaches is that LOF can work with multidimensional data, in fact, it might work _better_ with multidimensional data.

![image](https://github.com/open-spaced-repetition/fsrs-optimizer/assets/32575846/0f01638b-52f1-4c57-bf52-c3e755b44f2e)

@Expertium, I have give it a try. The LOF predicts that all data are inliers. (I use @user1823's data)

**Comment by L-M-Sherlock** [comment]
I have another method. We can sort the `delta_t` by `count` ascend. Then we start remove `delta_t` from the first row and accumulate the `count` we have removed. Finally, we can determine a upper limit (percentage, e.g. 5%) for removing outliers and stop in that point.

**Comment by Expertium** [comment]
> The LOF predicts that all data are inliers.

I assume you let it choose the % of contamination. How about one last try - manually set contamination %, say, 5% or 10%.

If that doesn't work either, then let's use your new proposed method.

**Comment by user1823** [comment]
> I have another method. We can sort the `delta_t` by `count` ascend.

You can try this method. But, the problem is that it will always filter out some data.

However, some values at the end of the dataset might represent true values from natural variation. Filtering out such data might lead to underfitting.

If Expertium's idea of manually setting contamination % doesn't work, I think that we should simply use ln(delta_t) with IQR.

**Comment by L-M-Sherlock** [comment]
> I assume you let it choose the % of contamination. How about one last try - manually set contamination %, say, 5% or 10%.

<img width="924" alt="image" src="https://github.com/open-spaced-repetition/fsrs-optimizer/assets/32575846/41438374-416b-4200-afb9-b769261f7b68">

The more weird things happen. I think these rows shouldn't be removed.

**Comment by Expertium** [comment]
Alright, forget about it then.

**Comment by L-M-Sherlock** [comment]
> If Expertium's idea of manually setting contamination % doesn't work, I think that we should simply use ln(delta_t) with IQR.

OK. But we should pass this test. As I mentioned before, here is a extreme case for outliers: https://github.com/open-spaced-repetition/fsrs4anki/issues/282#issuecomment-1643316187

Here is his dataset for pretrain (with outliers): [stability_for_pretrain.csv](https://github.com/open-spaced-repetition/fsrs-optimizer/files/12328926/stability_for_pretrain.csv)

**Comment by user1823** [comment]
So, just verify whether using ln(delta_t) filters out the outliers well in this collection or not.

**Comment by L-M-Sherlock** [comment]
I find that ln(delta_t) filters tend to give loose threshold. For example, if the Q1 = 2 and Q3 = 10, the original threshold = Q3 + 1.5 * IQR = 10 + 1.5 * 8 = 22. When we apply ln(), Q1 = 0.69 and Q3 = 2.3, the new original threshold = 2.3 + 1.5 * 1.61 = 4.715. e ^ 4.715 = 111.61, which is too large.

**Comment by user1823** [comment]
So, let's try your method:
> We can sort the delta_t by count ascend. Then we start to remove delta_t from the first row one by one and accumulate the count we have removed. Finally, we can determine a upper limit (percentage, e.g. 5%) for removing outliers and stop in that point.

To reduce the risk of filtering out inliers, we can add the condition that if the first row (in the sorted data) contains more than 5% of the data, we won't remove any data.

**Comment by L-M-Sherlock** [comment]
![pretrain_3](https://github.com/open-spaced-repetition/fsrs-optimizer/assets/32575846/23323715-aa24-427c-8fef-ca2654398593)

It removed 140 outliers (4.47%).

**Comment by L-M-Sherlock** [comment]
```python
        def remove_outliers(group: pd.DataFrame) -> pd.DataFrame:
            grouped_group = group.groupby(by=['r_history', 'delta_t'], group_keys=False).agg({'y': ['mean', 'count']}).reset_index()
            sort_index = grouped_group.sort_values(by=[('y', 'count')], ascending=True).index

            total = sum(grouped_group[('y', 'count')])
            has_been_removed = 0
            for i in sort_index:
                count = grouped_group.loc[i, ('y', 'count')]
                if has_been_removed >= total * 0.05 or count >= total * 0.05:
                    break
                has_been_removed += count
            group = group[group['delta_t'].isin(grouped_group[grouped_group[('y', 'count')] >= count]['delta_t'])]
            return group
```

Here is the code.

**Comment by user1823** [comment]
In my collection, this method would remove
- 9 outliers (3.57%) for `Again`
<img width="220" alt="image" src="https://github.com/open-spaced-repetition/fsrs-optimizer/assets/92206575/b35ca850-0269-41e9-a17a-b90db82c964c">

- no outliers (0%) for `Hard`
<img width="215" alt="image" src="https://github.com/open-spaced-repetition/fsrs-optimizer/assets/92206575/87148c0f-6d83-4768-b772-491d487fbee7">

- 11 outliers (1.92%) for `Good`
<img width="227" alt="image" src="https://github.com/open-spaced-repetition/fsrs-optimizer/assets/92206575/c4a8b457-b5ad-4402-bdd0-0cb62d248c10">

- 3 outliers (2.3%) for `Easy`
<img width="217" alt="image" src="https://github.com/open-spaced-repetition/fsrs-optimizer/assets/92206575/67837b88-1b06-4a76-9293-edf84cdc8106">

I am quite satisfied with these results.

**Comment by L-M-Sherlock** [comment]
@Expertium could you test it in your collection?

**Comment by user1823** [comment]
> ```py
> if has_been_removed >= total * 0.05 or count >= total * 0.05:
> ```

This should be
```py
if has_been_removed >= total * 0.05 or has_been_removed + count >= total * 0.05:
```

**Comment by L-M-Sherlock** [comment]
```python
has_been_removed + count >= total * 0.05
```

This condition is enough.

**Comment by user1823** [comment]
> ```python
> has_been_removed + count >= total * 0.05
> ```
> 
> This condition is enough.

Oh, you are right. So, update the condition in your previous comment so that Expertium can try it out with the correct code.

**Comment by Expertium** [comment]
Where do I put this code if Sherlock stopped releasing "open" versions of the optimizer (where all the code is visible in google colab and you can edit it) ages ago?

**Comment by L-M-Sherlock** [comment]
You can still modify the code based on this version: https://github.com/open-spaced-repetition/fsrs4anki/blob/main/archive/candidate/outlier_filter.ipynb

**Comment by user1823** [comment]
Well, @L-M-Sherlock, this version is much behind the release version (especially in pre-train). So, it would be very difficult to accurately judge the impact of the current change.

So, I would suggest you to push the change to a separate branch and then tell us how to import that change into optimizer.ipynb (like you told us in https://github.com/open-spaced-repetition/fsrs-optimizer/issues/5#issuecomment-1656794013).

**Comment by L-M-Sherlock** [comment]
But this feat is only related to pre-train.

And I have mentioned that the feat is not relied on RMSE.

**Comment by user1823** [comment]
> But this feat is only related to pre-train.
> 
> And I have mentioned that the feat doesn't rely on RMSE.

Yes, but testing this in the optimizer is faster as compared to using Excel to find out which `delta_t` would be filtered. Also, testing it in the optimizer allows us to see the impact on the predicted stability.

**Comment by L-M-Sherlock** [comment]
```
%pip install git+https://github.com/open-spaced-repetition/fsrs-optimizer@Feat/new-outlier-filter-based-on-count
```

You can install the branch version of FSRS optimizer in your notebook with above command. Just replace this line:

![image](https://github.com/open-spaced-repetition/fsrs-optimizer/assets/32575846/8ff07727-8ab2-48e0-8769-9f8e5f89db2e)

**Comment by user1823** [comment]
So, my results are as follows:
| | IQR      | Counts |
| ----------- | ----------- | ----------- |
| S0 | {1: 1.10, 2: 1.34, 3: 9.73, 4: 36.55} | {1: 1.10, 2: 1.48, 3: 14.48, 4: 36.91} |
| Data | <img width="217" alt="image" src="https://github.com/open-spaced-repetition/fsrs-optimizer/assets/92206575/8298a22c-3924-4d17-8147-5aefc8b26331"> | <img width="194" alt="image" src="https://github.com/open-spaced-repetition/fsrs-optimizer/assets/92206575/6958cff7-b9b5-4f37-9360-92c281b698a1"> |

I am totally satisfied with the results. If @Expertium is also satisfied with the results, we can merge this.

**Comment by user1823** [comment]
@L-M-Sherlock, this is unrelated but quite minor to create a new issue for.

Use s0 = 1.5 for Hard. Currently, 0.6 is used, which is too small.

**Comment by L-M-Sherlock** [comment]
> Use s0 = 1.5 for Hard. Currently, 0.6 is used, which is too small.

Do you have any stats for that?

**Comment by Expertium** [comment]
Speakign of which, remember how I said that when I was running the benchmark on 66 collections, I also wrote down S0?
Here are the average values, weighted by ln(reviews):
S0(Again)=0.7
S0(Hard)=1.1
S0(Good)=3.4
S0(Easy)=12.3

**Comment by user1823** [comment]
> I suggest running a statistical significance test to determine whether these values are better than the ones currently used.

In my opinion, we should just replace the S0 for Hard because the currently used value for `Hard` doesn't make much sense.

Also, the result of such a change would not be statistically significant because it would only affect the values in those collections that have a very low number of reviews with `Hard` as the first rating. So, we don't need to run a statistical significance test here.

**Comment by Expertium** [comment]
I want Sherlok to replace all 4 values though.

**Comment by user1823** [comment]
@Expertium, please try out the new outlier filter approach so that we can merge the branch and close this issue.

The way to use it was described by Sherlock here: https://github.com/open-spaced-repetition/fsrs-optimizer/issues/16#issuecomment-1676281709

**Comment by Expertium** [comment]
Maybe I'm doing something wrong,but I get exactly the same values of initial stability with both versions.

**Comment by L-M-Sherlock** [comment]
> Maybe I'm doing something wrong,but I get exactly the same values of initial stability with both versions.

Could you check the forgetting curves generated in pre-train?

**Comment by user1823** [comment]
> Maybe I'm doing something wrong,but I get exactly the same values of initial stability with both versions.

You need to click "Delete and restart runtime" after you change the optimizer version and then upload the collection file again (but the timezone settings and everything else would be preserved).

**Comment by user1823** [comment]
In my opinion, the new approach is better at identifying outliers because it can even filter those rows that have low `count` (and thus unreliable R) but are located in the middle of the data.

So, even if it doesn't perform better for your collection, it is definitely not worse. So, it makes sense to implement this.

Also, @L-M-Sherlock, in Expertium's pretrain data, I noticed that there were several rows with the same `count`. So, we should sort by the `count` & then the `delta_t` such that larger `delta_t` are filtered before smaller `delta_t` if they have the same `count`.

**Comment by L-M-Sherlock** [comment]
> So, we should sort by the `count` & then the `delta_t` such that larger `delta_t` are filtered before smaller `delta_t` if they have the same `count`.

Done in commit: https://github.com/open-spaced-repetition/fsrs-optimizer/commit/9dec42b879e92f61d00efed720d652ee074b7acd

**Comment by user1823** [comment]
Then, just merge it.

**Comment by user1823** [comment]
Although we decided that we won't use RMSE to determine the effectiveness of this change, I wanted to highlight that this change decreased the RMSE even though the number of reviews increased, at least for Sherlock's collection.

<img width=250 src="https://github.com/open-spaced-repetition/fsrs-optimizer/assets/92206575/31736c65-dc86-4bcf-89e9-5a4253df68ae"><img width=250 src="https://github.com/open-spaced-repetition/fsrs-optimizer/assets/92206575/179de929-4f4d-4362-85c2-63470540819c">
